src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -580,7 +624,33 @@ ReactDOMComponent.Mixin = {
     }
 
     var mountImage;
-    if (transaction.useCreateElement) {
+    if (nodesToReuse) {
+      // Check to see if the reusable node is of the correct type. It should be
+      // a single element (not an array, which is used for text components), and
+      // it should be an ELEMENT node.
+      if (Array.isArray(nodesToReuse) || nodesToReuse.nodeType !== ELEMENT_NODE_TYPE) {
+        MarkupMismatchError.throwComponentTypeMismatchError(nodesToReuse,
+          `A <${this._currentElement.type}> element.`);
+      }
+
+      // Check to make sure the node to reuse is the same tag as this component.
+      if (this._currentElement.type !== nodesToReuse.tagName.toLowerCase()) {
+        MarkupMismatchError.throwNodeTypeMismatchError(
+          nodesToReuse, this._currentElement.type);
+      }
+      // Precache the node, like we do in the useCreateElement branch, for event
+      // handling.
+      ReactDOMComponentTree.precacheNode(this, nodesToReuse);
+      this._flags |= Flags.hasCachedChildNodes;
```

Comment:
This was blatantly copied from the `if (transaction.useCreateElement)` branch. It seems to work, but I'm not entirely sure what it does.